packages/contracts/contracts/abstracts/libs/logic/AtomicSwapStateLogic.sol

```diff
@@ -59,12 +62,444 @@ library AtomicSwapStateLogic {
+            revert IAtomicSwapBase.BidNotInPlacedStatus(selectedBid.status);
+        }
+        // Update the bid status to 'Cancelled'
+        selectedBid.status = IAtomicSwapBase.BidStatus.Cancelled;
+
+        // Retrieve the buy tokens details from storage
+        IAtomicSwapBase.Coin storage buyToken = swapOrder[orderID].buyToken;
+
+        // If buy token is an ERC20 token, transfer it back to the bidder
+        if (buyToken.token != address(0)) {
+            TransferHelper.safeTransfer(
+                buyToken.token,
+                selectedBid.bidder,
+                selectedBid.amount
+            );
+        } else {
+            // If buy token is Ether, transfer it back to the bidder
+            TransferHelper.safeTransferETH(
+                msg.sender,
+                selectedBid.amount
+            );
+        }
+        return selectedBid.bidder;
+    }
+
+    function executeSwapWithPermit(
+        mapping( bytes32 => IAtomicSwapBase.AtomicSwapOrder
+    ) storage swapOrder,
+        IAtomicSwapBase.SwapWithPermitMsg calldata swap,
+        IAtomicSwapBase.Release[] calldata releases,  
+        IVesting vestingManager, 
+        IAtomicSwapBase.FeeParams memory feeParams
+        // uint    sellerFeeRate,
+        // uint    buyerFeeRate,
+        // uint    MAX_FEE_RATE_SCALE,
+        // address treasury
+    ) external returns (bytes32 id, address, address) {
+        id = generateNewAtomicSwapID(swap.uuid, address(this));
+        if(swapOrder[id].maker != address(0)) {
+            revert IAtomicSwapBase.OrderAlreadyExists();
+        }
+        if(swap.sellToken.token == address(0) || swap.buyToken.token == address(0)) {
+            revert IAtomicSwapBase.UnsupportedTokenPair();
+        }
+        
+        IAtomicSwapBase.PermitSignature memory makerSignature = parseSignature(
+            swap.makerPermitSignature
+        );
+        if(makerSignature.deadline > block.timestamp) {
+            revert IAtomicSwapBase.InvalidExpirationTime(makerSignature.deadline, block.timestamp);
+        }
+
+        IAtomicSwapBase.PermitSignature memory takerSignature = parseSignature(
+            swap.takerPermitSignature
+        );
+        if(takerSignature.deadline > block.timestamp) {
+            revert IAtomicSwapBase.InvalidExpirationTime(takerSignature.deadline, block.timestamp);
+        }
+
+        // Call permit to approve token transfer
+        IERC20Permit(swap.sellToken.token).permit(
+            makerSignature.sender,
+            address(this),
+            swap.sellToken.amount,
+            makerSignature.deadline,
+            makerSignature.v, makerSignature.r, makerSignature.s
+        );
+
+        // // (release)
+        IERC20Permit(swap.buyToken.token).permit(
+            takerSignature.sender,
+            address(this),
+            swap.sellToken.amount,
+            takerSignature.deadline,
+            takerSignature.v, takerSignature.r, takerSignature.s
+        );
```

Comment:
Do `buyToken` and `sellToken` have to implemented EIP 2612? or it can be all ERC20 tokens?